SVA/lib/secmem.c

```diff
@@ -523,6 +537,59 @@ sva_ghost_fault (uintptr_t vaddr) {
   struct CPUState * cpup = getCPUState();
   struct SVAThread * threadp = cpup->currentThread;
 
+  /* copy-on-write page fault */
+  if((code & PGEX_P) && (code & PGEX_W)){
+     printf("sva_ghost_fault COW enters\n");
+     pml4e_t * pml4e_ptr = get_pml4eVaddr (get_pagetable(), vaddr);
+     if(!isPresent (pml4e_ptr)) 
+        panic("sva_ghost_fault: cow pgfault pml4e %p does not exist\n", pml4e);
+     pdpte_t * pdpte = get_pdpteVaddr (pml4e_ptr, vaddr);
+     if(!isPresent (pdpte)) 
+        panic("sva_ghost_fault: cow pgfault pdpte %p does not exist\n", pdpte);
+     pde_t * pde = get_pdeVaddr (pdpte, vaddr);
+     if(!isPresent (pde)) 
+        panic("sva_ghost_fault: cow pgfault pde %p does not exist\n", pde);
+     pte_t * pte = get_pteVaddr (pde, vaddr);
+     uintptr_t paddr = *pte & PG_FRAME;
+     page_desc_t * pgDesc = getPageDescPtr (paddr);
+
+     if(pgDesc->type != PG_GHOST)
+	panic("SVA: sva_ghost_fault: vaddr = 0x%lx paddr = 0x%lx is not a ghost memory page!\n", vaddr, paddr); 
+     /* If only one process maps this page, directly grant this process write permission */
+
+     unprotect_paging();
+     if(pgDesc->count == 1)
+     {
+        * pte = (* pte) | PTE_CANWRITE;
+     }
+     /* Otherwise copy-on-write */
+     else
+     {
+        uintptr_t vaddr_old = (uintptr_t) getVirtual(paddr);
+        uintptr_t paddr_new = alloc_frame();
+        page_desc_t * pgDesc_new = getPageDescPtr (paddr_new);
+        if (pgRefCount (pgDesc_new) > 1) {
+                panic ("SVA: Ghost page still in use somewhere else!\n");
+        }
+        if (isPTP(pgDesc_new) || isCodePG (pgDesc_new)) {
+                panic ("SVA: Ghost page has wrong type!\n");
+        }
+
+
+     	memcpy(getVirtual(paddr_new), (void *) vaddr_old, X86_PAGE_SIZE);   
```

Comment:
This line would not work since provideSVAMemory() has already removed the kernel's direct mapping for this page when it becomes a ghost page. The only way to deal with this problem is to port the SVA DMAP from SVAMPX/mmu-devel branch to  the main branch. Please let me know what you think.